### ISSUE TITLE
any timeline on base quality?

### ISSUE THREAD

**Issue by kishwarshafin** [nanoporetech/bonito#25]
Hey @iiSeymour ,

Internally we have been leaning toward using Bonito but the absence of base quality is halting a lot of integration.  
I think you can simply derive the Q values using the inputs of CTC and a linear transformation of aggregated Q values from CTC encoding? It won't be great but will give enough to start with. Do you know when would this be incorporated?

**Comment by iiSeymour** [comment]
Deepnano did add this recently actually https://github.com/fmfi-compbio/deepnano-blitz/commit/1da9c01ca57d152207127bf79a8126855b255785

We could add this if you think it's good enough for your needs.

**Comment by kishwarshafin** [comment]
I was trying to find out how exactly they did it but I think the code is somewhere in an external dependency. However, if you incorporate this, I can do the debugging part and provide Q-Q plots on HG002 and zymo, so you know how informative the read qualities are. At this moment anything would be helpful.

**Comment by TimD1** [comment]
Hi @kishwarshafin,
I saw that you needed quality scores with bonito, and I realized that I'll probably need that myself soon, so yesterday and this morning I implemented quality scores for the greedy CTC decoder. You can find my bonito branch [here](https://github.com/TimD1/bonito). Just add the `--fastq` flag (and `--qbias` and `--qscale` if desired) to `bonito basecaller`. I haven't tested it that thoroughly yet, so let me know if you find any bugs.

@iiSeymour I would issue a pull request, but I took a few shortcuts (assuming the CTC matrix is 2D, and not allowing beam-search decoding), which probably broke decoding for training. It looks like integrating quality scores into the beam search decoder won't be as simple.

Anyways, I tested my branch on a small dataset and everything seems to be working fine. The latter plot is basically a QQ plot, it shows the proportion of bases which were incorrect for each Qscore. The black line shows what the Qscores should be if they were numerically correct.
![qscore_dist](https://user-images.githubusercontent.com/13918078/79048024-3b8dcf80-7be8-11ea-8724-fe7b3d395ed0.png)
![p_error](https://user-images.githubusercontent.com/13918078/79048026-3e88c000-7be8-11ea-905b-46afb28c7633.png)

**Comment by iiSeymour** [comment]
Nice, thanks @TimD1.

I made a start on the beam search implementation last night. I'll look to get this supported fully over the next few days.

**Comment by kishwarshafin** [comment]
@TimD1 , fantastic! The Q-values look very informative in your version.

**Comment by MichelMoser** [comment]
Hi, I am super interested in the score implementation.
Is qscore-branch runnable for some testing?

**Comment by iiSeymour** [comment]
Yes, it's ready to go @MichelMoser please give it a try and let me know. 

To calibrate the scores just add the following section to the `bonito/models/dna_r9.4.1/config.toml`

```
[qscores]
bias = 2.0
scale = 0.7
```

Adding the calibration section manually is only required for testing the branch. After I merge and release, the config will contain the section.

**Comment by MichelMoser** [comment]
great, thank you very much. 
Will launch it right away.

**Comment by MichelMoser** [comment]
Do I have to specify weights and models manually? 

`bonito basecaller dna_r9.4.1 --fastq --half tmp_f/${FNAME}/ ` throws:  

```
> loading model
Traceback (most recent call last):
  File "/usr/local/bin/bonito", line 9, in <module>
    load_entry_point('ont-bonito', 'console_scripts', 'bonito')()
  File "/opt/bonito/bonito/__init__.py", line 39, in main
    args.func(args)
  File "/opt/bonito/bonito/basecaller.py", line 19, in main
    model = load_model(args.model_directory, args.device, weights=int(args.weights), half=args.half)
  File "/opt/bonito/bonito/util.py", line 205, in load_model
    raise FileNotFoundError("no model weights found in '%s'" % dirname)
FileNotFoundError: no model weights found in 'dna_r9.4.1'

```

**Comment by iiSeymour** [comment]
When using bonito from GitHub you need to download the models - `bonito download --models`.

**Comment by iiSeymour** [comment]
I will merge and release tomorrow unless anybody has any objections/suggestions?

**Comment by TimD1** [comment]
Looks good to me, thanks for integrating it so quickly!

**Comment by kishwarshafin** [comment]
@iiSeymour ,

Thanks, so far I haven't had any issues with this. Please merge, hopefully, will be able to dive into this after London calling.

**Comment by MichelMoser** [comment]
Ran it without a problem, thank you!
 
Just had a look at some old nanopore data (December 2018) where bonito quality scores often did not reach guppy scores. But that might be a problem with the model as older firmware had different squiggle profiles, right?
I assume bonitos dna_r9.4.1 was trained on data from more recent firmware. 
Will rebasecall a flowcell from 2020 overnight.

![image](https://user-images.githubusercontent.com/4695665/80146001-66493200-85b1-11ea-98b0-be3e0a249684.png)

**Comment by iiSeymour** [comment]
Look likes it needs a bit of recalibration, see how the new data looks tomorrow but I'd up the bias by a few points.

**Comment by iiSeymour** [comment]
I have released [v0.1.4](https://github.com/nanoporetech/bonito/releases/tag/v0.1.4) with fastq output support to PyPI.
